lib/spack/spack/test/url_fetch.py

```
@@ -342,3 +342,79 @@ def _which(*args, **kwargs):
                 out = stage.fetch()
             assert err_fmt.format("curl") in out
+
+
+def test_url_fetch_text_without_url(tmpdir):
+    with pytest.raises(web_util.FetchError, match="URL is required"):
+        web_util.fetch_url_text(None)
+
+
+def test_url_missing_curl(tmpdir, monkeypatch):
+    """Check if URL's curl is missing."""
+    err_fmt = "No such command {0}"
+
+    def _which(*args, **kwargs):
+        err_msg = err_fmt.format(args[0])
+        raise spack.util.executable.CommandNotFoundError(err_msg)
+
+    # Patching the 'which' symbol imported by spack.util.web needed due
+    # to 'from spack.util.executable import which' in this module.
+    monkeypatch.setattr(spack.util.web, "which", _which)
+
+    with spack.config.override("config:url_fetch_method", "curl"):
+        with pytest.raises(web_util.FetchError, match="Missing required fetch method"):
+            web_util.fetch_url_text("https://github.com/")
+
+
+def test_url_check_curl_errors():
+    """Check that standard curl error returncodes raise expected errors."""
+    # Check returncode 22 (i.e., 404)
+    with pytest.raises(web_util.FetchError, match="not found"):
+        web_util.check_curl_code(22)
+
+    # Check returncode 60 (certificate error)
+    with pytest.raises(web_util.FetchError, match="invalid certificate"):
+        web_util.check_curl_code(60)
+
+
+def test_url_fetch_text_curl_failures(tmpdir, monkeypatch):
+    """Check failures if URL's curl is missing."""
+    err_fmt = "No such command {0}"
+
+    def _which(*args, **kwargs):
+        err_msg = err_fmt.format(args[0])
+        raise spack.util.executable.CommandNotFoundError(err_msg)
+
+    # Patching the 'which' symbol imported by spack.util.web needed due
+    # to 'from spack.util.executable import which' in this module.
+    monkeypatch.setattr(spack.util.web, "which", _which)
+
+    with spack.config.override("config:url_fetch_method", "curl"):
+        exists = web_util.url_exists("https://github.com/")
```

My concern here is the same as https://github.com/spack/spack/pull/29026#discussion_r933505285: if we require `curl`, and it isn't there, we should raise an exception vs. saying the URL doesn't exist.

---

Checking the new exception.